### PR TITLE
3DVar

### PR DESCRIPTION
**Pull request by AndreyAPopov**

### CODE REVIEW

File: src/+datools/+variational/ThreeDVar.m [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+            t = obj.Model.TimeSpan(end);
+
+            H = @(x) obs.observeWithoutError(t, x);
+            Hadjoint = @(x) obs.linearization(t, x).';
+            
+            J = @(x) cost(x, xb, dB, y, dR, H, Hadjoint);
+
+            opts = optimoptions('fmincon','Display','none','SpecifyObjectiveGradient',true, ...
+                'HessianApproximation', {'lbfgs', 50});
```
Comment (amitns122) [inline src/+datools/+variational/ThreeDVar.m:83]: We can specify Hessian vector products. Look at HessianMultiplyFcn option for fmincon.

Comment (amitns122) [reply src/+datools/+variational/ThreeDVar.m:83]: define hess = @(Hinfo, v) hessfunc(Hinfo, v, ........ stuff to compute the HVP)

Comment (amitns122) [reply src/+datools/+variational/ThreeDVar.m:83]: options = optimoptions('fmincon', 'SpecifyObjectiveGradient', true, 'CheckGradients', false, ...
    'Algorithm', 'trust-region-reflective', 'HessianMultiplyFcn', hess, 'Display', 'off');

File: src/+datools/+variational/ThreeDVar.m [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            xa = fmincon(J, xb, [],[],[],[],[],[],[], opts);
+
+            obj.State = xa;
+
+            function [c, g] = cost(x, xb, dB, y, dR, H, Hadjoint)
```
Comment (amitns122) [inline src/+datools/+variational/ThreeDVar.m:88]: function [c, g, Hinfo] = ...

File: src/+datools/+variational/ThreeDVar.m [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+                Bix = dB\(x - xb);
+
+                Riy = dR\(Hx - y);
+                
+                c = 0.5*(x - xb).'*Bix + 0.5*(Hx - y).'*Riy;
+
+                g = Bix + Hadjoint(x)*Riy;
+                
+            end
```
Comment (amitns122) [inline src/+datools/+variational/ThreeDVar.m:100]: Hinfo = [ ]. 
Can return stuff but matlab does not allow a direct hvp output, because its weird.

Comment (amitns122) [reply src/+datools/+variational/ThreeDVar.m:100]: function hv =hessfunc(hinfo, v, dB, H, dR)
hv = dB\v + H.'*(dR\(H*v));
end

Just use the linearization and stuff as well.

**Review by amitns122 — CHANGES_REQUESTED** [review]

**Comment by AndreyAPopov** [comment]
The reason why I don't want to add Hessian information is because with Hesian vector products the method would have to use a linear solver. For large systems this could be significantly more expensive than using LBFGS. I vote we simply use LBFGS and be done with it.

**Comment by AndreyAPopov** [comment]
I don't have a copy of matlab with the optimization toolbox installed, so can you finish this if possible? I don't want to leave us without even a simple 3DVar

**Comment by amitns122** [comment]
Alright. I'll download this version and finish working on it.

**Comment by amitns122** [comment]
@AndreyAPopov Okay. I think 3DVar is done.

File: src/+datools/+variational/ThreeDVar.m [hunk 4]
```diff
@@ -66,37 +71,60 @@ function forecast(obj)
...
-            opts = optimoptions('fmincon','Display','none','SpecifyObjectiveGradient',true, ...
-                'HessianApproximation', {'lbfgs', 50});
-            xa = fmincon(J, xb, [],[],[],[],[],[],[], opts);
+            if strcmp(obj.OptAlg, 'lbfgs')
+                opts = optimoptions('fmincon','Display','none', ...
+                    'SpecifyObjectiveGradient',true, ...
+                    'HessianApproximation', {'lbfgs', 30});
+            else
+
+                hessv = @(Hax, v) dB\v + Hax*(dR\(Hax.'*v));
```
Comment (AndreyAPopov) [inline src/+datools/+variational/ThreeDVar.m:102]: I think technically this needs to be Hax' not Hax.', as we want the adjoint not the transpose.

File: src/+datools/+variational/ThreeDVar.m [hunk 5]
```diff
@@ -66,37 +71,60 @@ function forecast(obj)
...
+                dR = decomposition(R, 'chol');
+            end
+
             xb = obj.State;
             obs = obj.Observation;
 
             t = obj.Model.TimeSpan(end);
 
             H = @(x) obs.observeWithoutError(t, x);
             Hadjoint = @(x) obs.linearization(t, x).';
```
Comment (AndreyAPopov) [inline src/+datools/+variational/ThreeDVar.m:92]: Ha, I make the same mistake here. I am so used to typing .'

File: src/+datools/+variational/ThreeDVar.m [hunk 6]
```diff
@@ -66,37 +71,60 @@ function forecast(obj)
...
+
             J = @(x) cost(x, xb, dB, y, dR, H, Hadjoint);
 
-            opts = optimoptions('fmincon','Display','none','SpecifyObjectiveGradient',true, ...
-                'HessianApproximation', {'lbfgs', 50});
-            xa = fmincon(J, xb, [],[],[],[],[],[],[], opts);
+            if strcmp(obj.OptAlg, 'lbfgs')
+                opts = optimoptions('fmincon','Display','none', ...
+                    'SpecifyObjectiveGradient',true, ...
+                    'HessianApproximation', {'lbfgs', 30});
```
[the remark this answers is not in the document]
Comment (amitns122) [reply src/+datools/+variational/ThreeDVar.m:99]: Changed to 50.  We can make it a parameter when DATools 2.0 is done.

**Pull request merged**